acvm-repo/acir/src/circuit/opcodes.rs

```diff
@@ -14,23 +14,85 @@ pub use memory_operation::{BlockId, MemOp};
 #[allow(clippy::large_enum_variant)]
 #[derive(Clone, PartialEq, Eq, Serialize, Deserialize)]
 pub enum Opcode {
+    /// An `AssertZero` opcode adds the constraint that `P(w) = 0`, where
+    /// `w=(w_1,..w_n)` is a tuple of `n` witnesses, and `P` is a multi-variate
+    /// polynomial of total degree at most `2`.
+    ///
+    /// The coefficients `{q_M}_{i,j}, q_i,q_c` of the polynomial are known
+    /// values which define the opcode.
+    ///
+    /// A general expression of assert-zero opcode is the following:
+    /// ```
+    /// \sum_{i,j} {q_M}_{i,j}w_iw_j + \sum_i q_iw_i +q_c = 0
+    /// ```
+    ///
+    /// An assert-zero opcode can be used to:
+    /// - **express a constraint** on witnesses; for instance to express that a
+    ///   witness `w` is a boolean, you can add the opcode: `w*w-w=0`
+    /// - or, to **compute the value** of an arithmetic operation of some inputs.
+    /// For instance, to multiply two witnesses `x` and `y`, you would use the
+    /// opcode `z-x*y=0`, which would constrain `z` to be `x*y`.
+    ///
+    /// The solver expects that at most one witness is not known when executing the opcode.
     AssertZero(Expression),
-    /// Calls to "gadgets" which rely on backends implementing support for specialized constraints.
+
+    /// Calls to "gadgets" which rely on backends implementing support for
+    /// specialized constraints.
     ///
-    /// Often used for exposing more efficient implementations of SNARK-unfriendly computations.  
+    /// Often used for exposing more efficient implementations of
+    /// SNARK-unfriendly computations.  
+    ///
+    /// All black box functions take as input a tuple `(witness, num_bits)`,
+    /// where `num_bits` is a constant representing the bit size of the input
+    /// witness, and they have one or several witnesses as output.
+    ///
+    /// Some more advanced computations assume that the proving system has an
+    /// 'embedded curve'. It is a curve that cycles with the main curve of the
+    /// proving system, i.e the scalar field of the embedded curve is the base
+    /// field of the main one, and vice-versa. The curves used by the proving
+    /// system are dependent on the proving system (and/or its configuration).
```

Comment:
```suggestion
    /// field of the main one, and vice-versa.
```
This feels slightly redundant no? Especially as we already state "Calls to "gadgets" which rely on backends implementing support" above